# EDA

В README не объясняют, что за задачу мы решаем и какие у нас данные, надо разобраться.


* В файле `problem_train.csv` в каждой строке содержится информация об объекте
с уникальным идентификатором, заданным в столбце `id`.

* В файле `problem_labels.csv` для каждого `id` из файла `problem_train.csv` приведена информация о
принадлежности к 14 категориям.

* В файле `problem_test.csv` дана информация об объектах, аналогичная содержащейся в файле
`problem_train.csv`.

* Используя `problem_train.csv` и `problem_labels.csv` в качестве данных для обучения, постройте файл
`problem_test_labels.csv`, в котором для каждого объекта из файла `problem_test.csv` укажите
вероятность его принадлежности к каждой из 14 рассматриваемых категорий.

* Для оценки качества полученного результата будет использоваться метрика `LogLoss`. По каждой из 14 
категорий метрика считается независимо, затем берется среднее арифметическое значений метрик,
полученных для категорий.

* Пожалуйста, предоставьте краткое описание пути поиска решения и код, использованный для получения
результата.

In [1]:
from IPython.core.display import HTML

HTML(
    r"""
<style>
    .output-plaintext, .output-stream, .output {
        font-family: "JetBrainsMono Nerd Font Mono"; # Any monospaced font should work
    }
</style>
"""
)

In [2]:
import os
from pathlib import Path

import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
import plotly.io as pio
import pandas as pd

In [3]:
pio.templates.default = "plotly_dark"

## TODO: change to white before submitting

### Откроем данные

In [4]:
train_df = pd.read_csv("./data/raw/problem_train.csv", low_memory=False)

In [5]:
train_df

id release  n_0000    n_0001    n_0002  n_0003  n_0004    n_0005  \
0     11193       a     NaN       NaN  0.025449     NaN     NaN  0.368421   
1     11382       a     NaN       NaN  0.031297     NaN     NaN  0.315789   
2     16531       a     NaN       NaN  0.024475     NaN     NaN  0.342105   
3      1896       a     NaN       NaN  0.041694     NaN     NaN  0.447368   
4     18262       c     NaN       NaN  0.038120     NaN     NaN  0.315789   
...     ...     ...     ...       ...       ...     ...     ...       ...   
7995  10898       a     NaN       NaN  0.053931     NaN     NaN  0.394737   
7996  16664       a     NaN       NaN  0.031731     NaN     NaN  0.394737   
7997   5334       c     NaN  0.904762  0.033463     NaN     NaN  0.394737   
7998   7905       c     NaN       NaN  0.047109     NaN     NaN  0.289474   
7999   3997       c     NaN  0.761905  0.049491     NaN     NaN  0.631579   

      n_0006  n_0007  ...  c_1368  c_1369  c_1370  c_1371  c_1372  c_1373  \
0        NaN     NaN  ...     NaN     NaN     NaN     NaN       a     NaN   
1        NaN     NaN  ...     NaN     NaN       a     NaN       a     NaN   
2        NaN     NaN  ...     NaN     NaN       a     NaN       a     NaN   
3        NaN     NaN  ...     NaN     NaN     NaN     NaN       a     NaN   
4        NaN     NaN  ...     NaN     NaN       b     NaN       a     NaN   
...      ...     ...  ...     ...     ...     ...     ...     ...     ...   
7995     NaN     NaN  ...     NaN     NaN     NaN     NaN       a     NaN   
7996     NaN     NaN  ...     NaN     NaN     NaN     NaN       a     NaN   
7997     NaN     NaN  ...     NaN     NaN     NaN     NaN       a     NaN   
7998     NaN     NaN  ...     NaN     NaN     NaN     NaN       a     NaN   
7999     0.0     0.0  ...     NaN     NaN     NaN     NaN       b     NaN   

      c_1374  c_1375  c_1376  c_1377  
0          q     NaN     NaN     NaN  
1        NaN     NaN     NaN     NaN  
2          b     NaN     NaN     NaN  
3        NaN     NaN     NaN     NaN  
4          a     NaN     NaN     NaN  
...      ...     ...     ...     ...  
7995     NaN     NaN     NaN       b  
7996       a     NaN     NaN     NaN  
7997     NaN     NaN     NaN     NaN  
7998       q     NaN     NaN     NaN  
7999     NaN     NaN     NaN     NaN  

[8000 rows x 1379 columns]

In [6]:
desc = train_df.describe(include="all")
desc

id release     n_0000      n_0001       n_0002      n_0003  \
count    8000.000000    8000  12.000000  388.000000  7662.000000  112.000000   
unique           NaN       3        NaN         NaN          NaN         NaN   
top              NaN       a        NaN         NaN          NaN         NaN   
freq             NaN    3838        NaN         NaN          NaN         NaN   
mean     9220.303375     NaN   0.307692    0.655744     0.040935    0.102749   
std      5236.897900     NaN   0.245454    0.216399     0.024599    0.153007   
min         0.000000     NaN   0.000000    0.000000     0.000000    0.000000   
25%      4773.500000     NaN   0.134615    0.428571     0.031974    0.011905   
50%      9221.500000     NaN   0.307692    0.702381     0.041369    0.047619   
75%     13702.750000     NaN   0.403846    0.857143     0.048950    0.142857   
max     18302.000000     NaN   0.846154    1.000000     1.000000    1.000000   

           n_0004       n_0005       n_0006      n_0007  ...  c_1368  c_1369  \
count   58.000000  7658.000000  1348.000000  995.000000  ...       3      48   
unique        NaN          NaN          NaN         NaN  ...       3      11   
top           NaN          NaN          NaN         NaN  ...       d       m   
freq          NaN          NaN          NaN         NaN  ...       1      22   
mean     0.352874     0.395981     0.193175    0.012012  ...     NaN     NaN   
std      0.209337     0.085000     0.236683    0.041382  ...     NaN     NaN   
min      0.066667     0.000000     0.000000    0.000000  ...     NaN     NaN   
25%      0.200000     0.342105     0.000000    0.000000  ...     NaN     NaN   
50%      0.283333     0.394737     0.200000    0.000000  ...     NaN     NaN   
75%      0.525000     0.421053     0.400000    0.011905  ...     NaN     NaN   
max      0.833333     1.000000     1.000000    1.000000  ...     NaN     NaN   

        c_1370  c_1371  c_1372  c_1373  c_1374  c_1375  c_1376  c_1377  
count      377       2    7998     440    3155     563       7    1908  
unique       3       2       2       5      24      19       2       2  
top          b       c       a       c       b       j       b       b  
freq       312       1    7460     167    1559     232       4    1441  
mean       NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN  
std        NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN  
min        NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN  
25%        NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN  
50%        NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN  
75%        NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN  
max        NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN  

[11 rows x 1379 columns]

#### Удостоверимся, что id не повторяется

In [7]:
train_df.id.duplicated().any()

False

#### Как определены фичи

In [8]:
px.histogram(desc.loc["count"])

In [9]:
na_cols = desc.loc["count"][desc.loc["count"] == 0].index
na_cols

Index(['n_0101', 'o_0224', 'o_0257', 'o_0308', 'c_0491', 'c_0529', 'c_0541',
       'c_0596', 'c_0632', 'c_0644', 'c_0693', 'c_0708', 'c_0710', 'c_0763',
       'c_0784', 'c_0803', 'c_0848', 'c_0902', 'c_0959', 'c_1009', 'c_1072',
       'c_1129', 'c_1148', 'c_1168', 'c_1230', 'c_1257', 'c_1277', 'c_1281',
       'c_1308', 'c_1332'],
      dtype='object')

### Посмотрим, какие размерности у категориальных фичей

In [10]:
px.histogram(desc.loc["unique"])

## TODO: Check if there are columns with text type. like: dog

### Посмотрим на типы данных

In [11]:
train_df.dtypes

id           int64
release     object
n_0000     float64
n_0001     float64
n_0002     float64
            ...   
c_1373      object
c_1374      object
c_1375      object
c_1376      object
c_1377      object
Length: 1379, dtype: object

In [12]:
train_df.dtypes.value_counts()

object     1025
float64     345
int64         9
dtype: int64

In [13]:
train_df.columns

Index(['id', 'release', 'n_0000', 'n_0001', 'n_0002', 'n_0003', 'n_0004',
       'n_0005', 'n_0006', 'n_0007',
       ...
       'c_1368', 'c_1369', 'c_1370', 'c_1371', 'c_1372', 'c_1373', 'c_1374',
       'c_1375', 'c_1376', 'c_1377'],
      dtype='object', length=1379)

In [14]:
train_df.columns.to_series().apply(lambda name: name[:2]).value_counts()

c_    1050
o_     211
n_     116
id       1
re       1
dtype: int64

#### Есть колонки начинающиеся с `o_`, это что?

In [15]:
train_df.columns[
    train_df.columns.to_series().apply(lambda name: str.startswith(name, "o_"))
]

Index(['o_0116', 'o_0117', 'o_0118', 'o_0119', 'o_0120', 'o_0121', 'o_0122',
       'o_0123', 'o_0124', 'o_0125',
       ...
       'o_0317', 'o_0318', 'o_0319', 'o_0320', 'o_0321', 'o_0322', 'o_0323',
       'o_0324', 'o_0325', 'o_0326'],
      dtype='object', length=211)

In [16]:
train_df["o_0326"]

0        NaN
1        NaN
2        NaN
3        NaN
4        NaN
        ... 
7995    11.0
7996     NaN
7997     NaN
7998     9.0
7999     NaN
Name: o_0326, Length: 8000, dtype: float64

#### Интересно, int64 -- не только id. Возможно, это те же n_, которые просто определились как int?

In [17]:
int_columns = train_df.dtypes[
    train_df.dtypes.apply(lambda dtype: dtype == np.int64)
].index
int_columns

Index(['id', 'n_0047', 'n_0050', 'n_0052', 'n_0061', 'n_0075', 'n_0091',
       'o_0176', 'o_0264'],
      dtype='object')

In [18]:
train_df[int_columns]

id  n_0047  n_0050  n_0052  n_0061  n_0075  n_0091  o_0176  o_0264
0     11193       1       1       1       1       1       1     303       7
1     11382       1       1       1       1       1       1     293       7
2     16531       1       1       1       1       1       1     131       0
3      1896       1       1       1       1       1       1     113       4
4     18262       1       1       1       1       1       1     240       6
...     ...     ...     ...     ...     ...     ...     ...     ...     ...
7995  10898       1       1       1       1       1       1       4       5
7996  16664       1       1       1       1       1       1     207       5
7997   5334       1       1       1       1       1       1     369       8
7998   7905       1       1       1       1       1       1      69       4
7999   3997       1       1       1       1       1       1     125       4

[8000 rows x 9 columns]

In [19]:
train_df[["n_0047", "n_0050", "n_0052", "n_0061", "n_0075", "n_0091"]].value_counts()

n_0047  n_0050  n_0052  n_0061  n_0075  n_0091
1       1       1       1       1       1         8000
dtype: int64

Какие-то бесполезные колонка, все значения 1
## TODO: удалить эти колонки???

### Подгрузим лейблы

In [20]:
train_labels_df = pd.read_csv("./data/raw/problem_labels.csv")
train_labels_df

id  service_a  service_b  service_c  service_d  service_e  service_f  \
0     11193          1          1          0          0          0          0   
1     11382          0          0          0          0          0          0   
2     16531          0          0          0          0          0          0   
3      1896          0          0          0          1          0          0   
4     18262          0          0          0          1          1          0   
...     ...        ...        ...        ...        ...        ...        ...   
7995  10898          1          1          1          0          0          0   
7996  16664          0          0          0          0          0          0   
7997   5334          1          0          0          0          0          0   
7998   7905          1          1          1          0          0          0   
7999   3997          1          1          0          0          1          0   

      service_g  service_h  service_i  service_j  service_k  service_l  \
0             0          0          0          1          1          0   
1             0          0          0          1          1          0   
2             0          0          0          1          1          0   
3             0          0          0          1          0          1   
4             0          0          0          0          1          1   
...         ...        ...        ...        ...        ...        ...   
7995          1          0          0          1          1          1   
7996          0          0          0          1          1          0   
7997          0          0          0          1          1          0   
7998          0          0          0          1          1          0   
7999          0          1          0          1          0          0   

      service_m  service_n  
0             0          0  
1             0          0  
2             0          0  
3             0          0  
4             1          0  
...         ...        ...  
7995          1          1  
7996          0          0  
7997          0          0  
7998          0          0  
7999          0          0  

[8000 rows x 15 columns]

Проверим, что айдишники совпадают

In [21]:
(train_labels_df.id == train_df.id).all()

True

Пока что ничего не понимаю. есть категориальные, флоут переменные. Возможно, это что-то нагенерили из исходных данных.

В названиях есть release, service -- мб это что-то связанное с разными продуктами? или с разными сервисами внутри тинькофф?

Одной записи соответствует несколько лейблов, то есть у нас мультилейбл задача

#### мб стоит взглянуть на матрицу корреляции между лейблами?)))

In [22]:
px.imshow(
    train_labels_df.drop(columns=["id"]).corr(),
)

Соседние лейблы в целом сильнее скоррелированы. например i-m = 0.80, j-k = 0.53, f-g = 0.40, d-e = 0.42, a-b = 0.56, b-c = 0.4.

Есть еще h-i = 0.30, c-g = 0.27

Эта корреляция мб не будет мешать модели, если на каждый будет отдельная модель

#### Все лейблы определены на всех записях

In [23]:
train_labels_df.describe()

id    service_a    service_b    service_c    service_d  \
count   8000.000000  8000.000000  8000.000000  8000.000000  8000.000000   
mean    9220.303375     0.472125     0.328250     0.261625     0.017000   
std     5236.897900     0.499254     0.469606     0.439547     0.129279   
min        0.000000     0.000000     0.000000     0.000000     0.000000   
25%     4773.500000     0.000000     0.000000     0.000000     0.000000   
50%     9221.500000     0.000000     0.000000     0.000000     0.000000   
75%    13702.750000     1.000000     1.000000     1.000000     0.000000   
max    18302.000000     1.000000     1.000000     1.000000     1.000000   

         service_e    service_f    service_g    service_h    service_i  \
count  8000.000000  8000.000000  8000.000000  8000.000000  8000.000000   
mean      0.053875     0.029125     0.051875     0.297000     0.016125   
std       0.225785     0.168167     0.221788     0.456965     0.125964   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       0.000000     0.000000     0.000000     1.000000     0.000000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

         service_j    service_k  service_l    service_m    service_n  
count  8000.000000  8000.000000  8000.0000  8000.000000  8000.000000  
mean      0.848250     0.782000     0.1080     0.088625     0.178000  
std       0.358801     0.412913     0.3104     0.284219     0.382537  
min       0.000000     0.000000     0.0000     0.000000     0.000000  
25%       1.000000     1.000000     0.0000     0.000000     0.000000  
50%       1.000000     1.000000     0.0000     0.000000     0.000000  
75%       1.000000     1.000000     0.0000     0.000000     0.000000  
max       1.000000     1.000000     1.0000     1.000000     1.000000

Вот только вопрос - 8000 это много или мало 🤔 Думаю, достаточно

### Посмотрим на тест

In [24]:
test_df = pd.read_csv("./data/raw/problem_test.csv", low_memory=False)
test_df

id release  n_0000    n_0001    n_0002  n_0003  n_0004    n_0005  \
0      7957       a     NaN       NaN  0.029890     NaN     NaN  0.368421   
1       693       c     NaN  0.452381  0.032164     NaN     NaN  0.421053   
2     15558       c     NaN  0.547619  0.049166     NaN     NaN  0.500000   
3     15614       a     NaN       NaN  0.048733     NaN     NaN  0.394737   
4      8149       b     NaN       NaN  0.049924     NaN     NaN  0.368421   
...     ...     ...     ...       ...       ...     ...     ...       ...   
1995    200       b     NaN       NaN  0.014728     NaN     NaN  0.368421   
1996  10876       a     NaN       NaN  0.052415     NaN     NaN  0.578947   
1997   4035       a     NaN       NaN  0.038770     NaN     NaN  0.368421   
1998   3640       b     NaN       NaN  0.039420     NaN     NaN  0.552632   
1999    157       b     NaN       NaN  0.043535     NaN     NaN  0.421053   

      n_0006    n_0007  ...  c_1368  c_1369  c_1370  c_1371  c_1372  c_1373  \
0        NaN       NaN  ...     NaN     NaN     NaN     NaN       a     NaN   
1        0.0       NaN  ...     NaN     NaN     NaN     NaN       a     NaN   
2        0.2       NaN  ...     NaN     NaN     NaN     NaN       a       e   
3        NaN       NaN  ...     NaN     NaN     NaN     NaN       b     NaN   
4        NaN       NaN  ...     NaN     NaN     NaN     NaN       a     NaN   
...      ...       ...  ...     ...     ...     ...     ...     ...     ...   
1995     NaN  0.059524  ...     NaN     NaN     NaN     NaN       a     NaN   
1996     NaN  0.000000  ...     NaN     NaN     NaN     NaN       a     NaN   
1997     NaN       NaN  ...     NaN     NaN     NaN     NaN       a     NaN   
1998     NaN       NaN  ...     NaN     NaN     NaN     NaN       a     NaN   
1999     NaN       NaN  ...     NaN     NaN     NaN     NaN       a     NaN   

      c_1374  c_1375  c_1376  c_1377  
0          q     NaN     NaN     NaN  
1          x     NaN     NaN     NaN  
2          b     NaN     NaN     NaN  
3        NaN     NaN     NaN     NaN  
4        NaN     NaN     NaN       b  
...      ...     ...     ...     ...  
1995     NaN     NaN     NaN     NaN  
1996     NaN     NaN     NaN     NaN  
1997       b     NaN     NaN     NaN  
1998     NaN     NaN     NaN     NaN  
1999     NaN     NaN     NaN     NaN  

[2000 rows x 1379 columns]

In [25]:
test_desc = test_df.describe(include="all")
test_desc

id release    n_0000      n_0001       n_0002     n_0003  \
count    2000.000000    2000  2.000000  116.000000  1904.000000  20.000000   
unique           NaN       3       NaN         NaN          NaN        NaN   
top              NaN       a       NaN         NaN          NaN        NaN   
freq             NaN     940       NaN         NaN          NaN        NaN   
mean     9166.887500     NaN  0.384615    0.658046     0.041033   0.107341   
std      5322.849729     NaN  0.217571    0.208902     0.024645   0.166684   
min        23.000000     NaN  0.230769    0.000000     0.006714   0.000000   
25%      4502.500000     NaN  0.307692    0.452381     0.032164   0.025794   
50%      9236.000000     NaN  0.384615    0.678571     0.041585   0.063492   
75%     13802.250000     NaN  0.461538    0.857143     0.049085   0.105159   
max     18296.000000     NaN  0.538462    0.976190     1.000000   0.714286   

           n_0004       n_0005      n_0006      n_0007  ...  c_1368  c_1369  \
count   14.000000  1903.000000  373.000000  258.000000  ...     0.0      12   
unique        NaN          NaN         NaN         NaN  ...     NaN       7   
top           NaN          NaN         NaN         NaN  ...     NaN       m   
freq          NaN          NaN         NaN         NaN  ...     NaN       4   
mean     0.295238     0.397157    0.213405    0.009690  ...     NaN     NaN   
std      0.188043     0.084669    0.252157    0.018267  ...     NaN     NaN   
min      0.066667     0.026316    0.000000    0.000000  ...     NaN     NaN   
25%      0.141667     0.342105    0.000000    0.000000  ...     NaN     NaN   
50%      0.283333     0.394737    0.200000    0.000000  ...     NaN     NaN   
75%      0.391667     0.421053    0.400000    0.011905  ...     NaN     NaN   
max      0.700000     0.947368    1.000000    0.166667  ...     NaN     NaN   

        c_1370  c_1371  c_1372  c_1373  c_1374  c_1375  c_1376  c_1377  
count      109     0.0    1999     108     779     149       1     495  
unique       2     NaN       2       5      13      15       1       2  
top          b     NaN       a       a       b       j       a       b  
freq        91     NaN    1841      41     391      53       1     372  
mean       NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN  
std        NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN  
min        NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN  
25%        NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN  
50%        NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN  
75%        NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN  
max        NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN  

[11 rows x 1379 columns]

In [26]:
test_na_cols = test_desc.loc["count"][test_desc.loc["count"] == 0].index
test_na_cols

Index(['n_0008', 'n_0011', 'n_0055', 'n_0101', 'n_0111', 'o_0117', 'o_0121',
       'o_0122', 'o_0126', 'o_0149',
       ...
       'c_1288', 'c_1292', 'c_1308', 'c_1312', 'c_1332', 'c_1334', 'c_1349',
       'c_1360', 'c_1368', 'c_1371'],
      dtype='object', length=179)

Тут уже много фичей неопределено

#### Удостоверимся, что у нас одинаковые фичи там и там

In [27]:
train_df.columns.intersection(test_df.columns)

Index(['id', 'release', 'n_0000', 'n_0001', 'n_0002', 'n_0003', 'n_0004',
       'n_0005', 'n_0006', 'n_0007',
       ...
       'c_1368', 'c_1369', 'c_1370', 'c_1371', 'c_1372', 'c_1373', 'c_1374',
       'c_1375', 'c_1376', 'c_1377'],
      dtype='object', length=1379)

## PCA/t-SNE на хорошо определенных фичах (90%+)

In [28]:
train_counts = train_df.describe(include="all").loc["count"]
train_counts[train_counts > 7000]

id         8000.0
release      8000
n_0002     7662.0
n_0005     7658.0
n_0019     7538.0
            ...  
c_1259       8000
c_1286       7995
c_1316       7911
c_1348       8000
c_1372       7998
Name: count, Length: 91, dtype: object

In [29]:
well_done_df = train_df[train_counts[train_counts > 7000].index]
well_done_df

id release    n_0002    n_0005  n_0019    n_0038  n_0047  n_0050  \
0     11193       a  0.025449  0.368421     0.0  0.193548       1       1   
1     11382       a  0.031297  0.315789     0.0  0.177419       1       1   
2     16531       a  0.024475  0.342105     0.0  0.290323       1       1   
3      1896       a  0.041694  0.447368     0.0  0.370968       1       1   
4     18262       c  0.038120  0.315789     0.0  0.177419       1       1   
...     ...     ...       ...       ...     ...       ...     ...     ...   
7995  10898       a  0.053931  0.394737     0.0  0.209677       1       1   
7996  16664       a  0.031731  0.394737     0.0  0.290323       1       1   
7997   5334       c  0.033463  0.394737     0.0  0.177419       1       1   
7998   7905       c  0.047109  0.289474     0.2  0.177419       1       1   
7999   3997       c  0.049491  0.631579     0.0  0.516129       1       1   

      n_0052  n_0061  ...  c_1223  c_1227  c_1236  c_1244  c_1252  c_1259  \
0          1       1  ...       c       a       c       d       b       n   
1          1       1  ...       a       a       c       d       b       e   
2          1       1  ...       c       a       a       d       b       w   
3          1       1  ...       c       a       c       d       b       e   
4          1       1  ...       c       a       c       d       b       e   
...      ...     ...  ...     ...     ...     ...     ...     ...     ...   
7995       1       1  ...       c       b       c       a       b       i   
7996       1       1  ...       c       a       c       d       b       w   
7997       1       1  ...       c       a       c       d       b       n   
7998       1       1  ...       c       a       a       a       b       n   
7999       1       1  ...       c       a       c       d       b       n   

      c_1286  c_1316  c_1348  c_1372  
0          b       b       b       a  
1          b       b       b       a  
2          b       b       b       a  
3          b       a       b       a  
4          b       b       b       a  
...      ...     ...     ...     ...  
7995       a       a       c       a  
7996       b       b       b       a  
7997       b       b       b       a  
7998       b       b       b       a  
7999       b       b       c       b  

[8000 rows x 91 columns]

#### todo: maybe remove o_ and similar int features, because they're actually categorical???

In [30]:
well_done_df.describe()

id       n_0002       n_0005       n_0019       n_0038  \
count   8000.000000  7662.000000  7658.000000  7538.000000  7286.000000   
mean    9220.303375     0.040935     0.395981     0.020350     0.238586   
std     5236.897900     0.024599     0.085000     0.072958     0.079213   
min        0.000000     0.000000     0.000000     0.000000     0.032258   
25%     4773.500000     0.031974     0.342105     0.000000     0.193548   
50%     9221.500000     0.041369     0.394737     0.000000     0.209677   
75%    13702.750000     0.048950     0.421053     0.000000     0.258065   
max    18302.000000     1.000000     1.000000     1.000000     0.919355   

       n_0047  n_0050  n_0052  n_0061       n_0067  ...       n_0083  n_0091  \
count  8000.0  8000.0  8000.0  8000.0  8000.000000  ...  7992.000000  8000.0   
mean      1.0     1.0     1.0     1.0     0.562482  ...     0.664654     1.0   
std       0.0     0.0     0.0     0.0     0.292893  ...     0.414820     0.0   
min       1.0     1.0     1.0     1.0     0.000000  ...     0.000000     1.0   
25%       1.0     1.0     1.0     1.0     0.357143  ...     0.250000     1.0   
50%       1.0     1.0     1.0     1.0     0.571429  ...     1.000000     1.0   
75%       1.0     1.0     1.0     1.0     0.857143  ...     1.000000     1.0   
max       1.0     1.0     1.0     1.0     1.000000  ...     1.000000     1.0   

            n_0108       n_0109       o_0120       o_0144       o_0176  \
count  8000.000000  8000.000000  7974.000000  7974.000000  8000.000000   
mean      0.487258     0.080375     0.474918     7.467018   198.395500   
std       0.261672     0.084939     0.894713     1.670485   118.487761   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.266667     0.000000     0.000000     6.000000    92.000000   
50%       0.466667     0.062500     0.000000     7.000000   193.000000   
75%       0.700000     0.125000     1.000000     8.000000   315.000000   
max       1.000000     0.750000     6.000000    18.000000   369.000000   

            o_0201       o_0230       o_0264  
count  7867.000000  7931.000000  8000.000000  
mean      1.733443     1.364393     4.585875  
std       1.129209     1.044707     2.237912  
min       0.000000     0.000000     0.000000  
25%       1.000000     1.000000     4.000000  
50%       2.000000     1.000000     5.000000  
75%       2.000000     2.000000     6.000000  
max       4.000000     3.000000    10.000000  

[8 rows x 22 columns]

#### Start with "n_" columns

In [31]:
cols = well_done_df.columns.to_series()
n_cols = cols[cols.apply(lambda name: str.startswith(name, "n_"))].index

numeric_df = well_done_df[n_cols]
numeric_df

n_0002    n_0005  n_0019    n_0038  n_0047  n_0050  n_0052  n_0061  \
0     0.025449  0.368421     0.0  0.193548       1       1       1       1   
1     0.031297  0.315789     0.0  0.177419       1       1       1       1   
2     0.024475  0.342105     0.0  0.290323       1       1       1       1   
3     0.041694  0.447368     0.0  0.370968       1       1       1       1   
4     0.038120  0.315789     0.0  0.177419       1       1       1       1   
...        ...       ...     ...       ...     ...     ...     ...     ...   
7995  0.053931  0.394737     0.0  0.209677       1       1       1       1   
7996  0.031731  0.394737     0.0  0.290323       1       1       1       1   
7997  0.033463  0.394737     0.0  0.177419       1       1       1       1   
7998  0.047109  0.289474     0.2  0.177419       1       1       1       1   
7999  0.049491  0.631579     0.0  0.516129       1       1       1       1   

        n_0067  n_0075    n_0078    n_0083  n_0091    n_0108  n_0109  
0     0.928571       1  0.800000  1.000000       1  0.800000  0.1875  
1     0.928571       1  0.666667  0.000000       1  0.666667  0.1875  
2     0.428571       1  0.833333  1.000000       1  0.833333  0.1875  
3     0.571429       1  0.566667  0.833333       1  0.566667  0.1875  
4     0.928571       1  0.600000  0.666667       1  0.600000  0.2500  
...        ...     ...       ...       ...     ...       ...     ...  
7995  0.000000       1  0.166667  0.000000       1  0.166667  0.0625  
7996  0.714286       1  0.733333  0.083333       1  0.733333  0.1250  
7997  1.000000       1  0.833333  1.000000       1  0.833333  0.0625  
7998  0.357143       1  0.400000  1.000000       1  0.400000  0.1250  
7999  0.428571       1  0.700000  1.000000       1  0.700000  0.0625  

[8000 rows x 15 columns]

PCA wants imputing

In [32]:
import numpy as np
from sklearn.impute import SimpleImputer
imp = SimpleImputer(missing_values=np.nan, strategy='mean')
imp.fit(numeric_df)
imputed_df = imp.transform(numeric_df)

In [33]:
from sklearn.decomposition import PCA

pca = PCA(n_components=5)
comps = pca.fit_transform(imputed_df)

In [34]:
pca.explained_variance_ratio_

array([0.48430617, 0.29450547, 0.1668958 , 0.02291048, 0.01235706])

мы смогли 60 фичей переварить в 3 компоненты с почти полными покрытием

In [35]:
train_labels_df

id  service_a  service_b  service_c  service_d  service_e  service_f  \
0     11193          1          1          0          0          0          0   
1     11382          0          0          0          0          0          0   
2     16531          0          0          0          0          0          0   
3      1896          0          0          0          1          0          0   
4     18262          0          0          0          1          1          0   
...     ...        ...        ...        ...        ...        ...        ...   
7995  10898          1          1          1          0          0          0   
7996  16664          0          0          0          0          0          0   
7997   5334          1          0          0          0          0          0   
7998   7905          1          1          1          0          0          0   
7999   3997          1          1          0          0          1          0   

      service_g  service_h  service_i  service_j  service_k  service_l  \
0             0          0          0          1          1          0   
1             0          0          0          1          1          0   
2             0          0          0          1          1          0   
3             0          0          0          1          0          1   
4             0          0          0          0          1          1   
...         ...        ...        ...        ...        ...        ...   
7995          1          0          0          1          1          1   
7996          0          0          0          1          1          0   
7997          0          0          0          1          1          0   
7998          0          0          0          1          1          0   
7999          0          1          0          1          0          0   

      service_m  service_n  
0             0          0  
1             0          0  
2             0          0  
3             0          0  
4             1          0  
...         ...        ...  
7995          1          1  
7996          0          0  
7997          0          0  
7998          0          0  
7999          0          0  

[8000 rows x 15 columns]

In [36]:
px.scatter(x=comps[:, 0], y=comps[:, 1])

 Похоже на проекцию цилиндра -- плотные бока, пустая середина 🤔

In [37]:
px.scatter(x=comps[:, 0], y=comps[:, 1], color=train_labels_df["service_a"])

In [38]:
px.scatter(x=comps[:, 0], y=comps[:, 1], color=train_labels_df["service_d"])

Мы даже на этих 15 количественных признаках получаем какую-то разделенность!! задача в принципе решаема

#### 3DDDDDDDD

In [39]:
column = "service_a"
px.scatter_3d(x=comps[:, 0], y=comps[:, 1], z=comps[:, 2], color=train_labels_df[column])

### TODO: some n_cols as n_0005 have discreate values that repeats like 0.177419, 0.209677, 0.290323
можно посчитать их дискретизацию

#### TODO: 
некоторые фичи могут почти не давать никакой информации в целом, но для конкретного лейбла быть очень важны
например признак наличия слова "суицид" в переписке → в лейбл `сервис психотерапия`

это имеет смысл, потому что у нас есть лейблы, у которых ооочень мало единиц


#### TODO: для некоторых признаков можно постромть фичу `признак is nan`. мб это будет коррелировать с какими-то сервисами

## TODO: BASELINE - use pca 3 components

### TODO: baseline -- более 80 %, только категориальные, ...

### TODO: можно сделать простые преобразования над нумерик фичами, у которых сильно изменяющися масштаб (e-3, e+3, ...)
например логарифм

## TODO: как заполнять наны?

```
x   y   z
1   a   NaN
2   a   42
4   b   NAN
-2  NaN NaN
5   a   NaN
```

* построить линрег f(x, y) -> z. для z ~ 20 % не нан

* 15% + 5% = train + test

* использовать эту новую фичу уже в бустинге